cirq/vis/histogram.py

```diff
@@ -0,0 +1,150 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Any, Mapping, Optional, Sequence, Union, SupportsFloat
+
+import numpy as np
+from matplotlib import pyplot as plt
+
+
+def integrated_histogram(
+    data: Union[Sequence[SupportsFloat], Mapping[Any, SupportsFloat]],
+    ax: Optional[plt.Axes] = None,
+    *,
+    cdf_on_x: bool = False,
+    axis_label: str = '',
+    semilog: bool = True,
+    median_line: bool = True,
+    median_label: Optional[str] = 'median',
+    mean_line: bool = False,
+    mean_label: Optional[str] = 'mean',
+    hide_zero: bool = True,
+    title: Optional[str] = None,
+    **kwargs,
+) -> plt.Axes:
+    """Plot the integrated histogram for an array of data.
+
+    Suppose the input is a list of gate fidelities. The x-axis of the plot will
+    be gate fidelity, and the y-axis will be the probability that a random gate
+    fidelity from the list is less than the x-value. It will look something like
+    this
+
+    1.0
+    |              |
+    |           ___|
+    |           |
+    |       ____|
+    |      |
+    |      |
+    |_____|_______________
+    0.0
+
+    Another way of saying this is that we assume the probability distribution
+    function (pdf) of gate fidelities is a set of equally weighted delta
+    functions at each value in the list. Then, the "integrated histogram"
+    is the cumulative distribution function (cdf) for this pdf.
+
+    Args:
+        data: Data to histogram. If the data is a mapping, we histogram the
+            values. All nans will be removed.
+        ax: The axis to plot on. If None, we generate one.
+        cdf_on_x: If True, flip the axes compared the above example.
+        axis_label: Label for x axis.
+        semilog: If True, force the x-axis to be logarithmic.
+        median_line: If True, draw a vertical line on the median value.
+        median_label: If drawing median line, optional label for it.
+        mean_line: If True, draw a vertical line on the mean value.
+        mean_label: If drawing mean line, optional label for it.
+        **plot_options: Kwargs to forward to `ax.step()`. Some examples are
+            color: Color of the line.
+            linestyle: Linestyle to use for the plot.
+            lw: linewidth for integrated histogram
+            ms: marker size for a histogram trace
+            label: An optional label which can be used in a legend.
+
+
+    Returns:
+        The axis that was plotted on.
+    """
+    show_plot = not ax
+    if ax is None:
+        fig, ax = plt.subplots(1, 1)
+
+    if isinstance(data, Mapping):
+        data = list(data.values())
+
+    data = [d for d in data if not np.isnan(d)]
+    n = len(data)
+
+    if not hide_zero:
+        bin_values = np.linspace(0, 1, n + 1)
+        parameter_values = sorted(np.concatenate(([0], data)))
```

Comment:
Can you help me understand this? It looks a bit fishy to me - we are modifying the data to show where zero is? Why not just modify the x axis min value?

Reply:
A step plot starts by plotting a horizontal line between the first (x0, y0) and (x1, y0) before going up from (x1, y0) to (x1, y1). Because of this, the first horizontal line would co-inside with x-axis (y values are generated by `np.linespace` so first `y` value would always be 0). 

If `hide_zero` is false, we move the plot up by 1 unit. For example, the left plot below is generated using `hide_zero = True` and right plot is generated using `hide_zero = False`. 
- Left: `cirq.integrated_histogram(range(1, 10), marker='o', ms=10.0, hide_zero=True)`
- Right: `cirq.integrated_histogram(range(1, 10), marker='o', ms=10.0, hide_zero=False)`
![image](https://user-images.githubusercontent.com/7863287/111854660-5a432d00-8946-11eb-8ace-4f12465aa86a.png)